### ISSUE TITLE
game-sims4: Things to improve support

### ISSUE THREAD

**Issue by fantesykikachu** [Nexus-Mods/Vortex#4928]
general quality of life improvements for The Sims 4 extension (game-sims4)

- [x] Don't use sub-directories: the sims 4 expects all .package and .ts4script files to be in the top level of the mods folder. now this would be fine if most mod developers for the sims 4 expected users to use a mod manager, but they don't.

- [x] Support for Tray items: basically all files with other extensions shouldn't go in "The Sims 4\Mods", but should instead go in "The Sims 4\Tray" (again no sub-directories), the extensions I know of are: .bpi, .blueprint,  .trayitem, .sfx, .ion, .householdbinary, .sgi, .hhi, .room, .midi, .rmi.

- [x]  Support for adding loose files to vortex: sometimes mod authors on distribute .package and .ts4script files, and it would be nice to add those to vortex without having to compress them first.

- [ ] Support for loose script files (HARD): the sims 4 supports python scripting, this is in the form of: .py, .pyc, and .pyo (no longer supported) files. however they must be either in a .zip file (not .7z or .rar) or in a .ts4script file (which is really just .zip file renamed. implementing support for this would be pretty tough so maybe just give a warning telling them to put all : .py and .pyc file in a .zip and rename it to .ts4script (and an error that the sims 4 no longer supports .pyo files)

**Comment by IDCs** [comment]
Hi @fantesykikachu, thank you for sending us your suggestion - this helps figuring out old and new installation patterns for our supported games. 

Regarding the first item you mentioned. As far as I can remember, not ALL *.package and *.ts4script mods will function without a sub-directory; simplest example (and lets be honest, the most used Sims 4 mod of all) Wicked Whims will not function correctly if its files are simply dropped inside the top level of the mods folder. 

Given that there is no way for us to ascertain where such mods are meant to be deployed programatically, I think I will add a new mod type to the extension which the users can select manually through the mods pane which searches for any .package and .ts4script files and deploys them to the top level of the mods directory. Are you aware of any other files or file extensions which are supposed to be deployed alongside .package/.ts4script files or is it safe for us to assume that the entirety of the mod is included within those files ?

**Comment by TanninOne** [comment]
There is a file called "resource.cfg" in the mods directory with this content:
```
Priority 500
PackedFile *.package
PackedFile */*.package
PackedFile */*/*.package
PackedFile */*/*/*.package
PackedFile */*/*/*/*.package
PackedFile */*/*/*/*/*.package
```

That would indicate that this file controls where the game searches for package files, in this case it tells it to search in nested dirs up to 5 levels deep.
If that is the case, other users may not have this file or have it with different content which could then cause mods to not work that work fine for others.

Now I don't know if/how this affects .ts4script files but I know for a fact that on this test system I have mods that have a ts4script file in a nested directory and they work in game, so it would seem this should work too.

I don't really believe that it makes a difference mod-to-mod, either the game reads ts4script/package files from nested directories or it doesn't, I doubt some mods _need_ to be installed into the base directory and others _need_ to be in subdirectories.

It may be prudent for us to have Vortex verify/write that resources.cfg file to ensure it works.

Mods that contain .py/.pyc files however apparently need to be installed into the mods base directory so it may make sense to handle those with a separate mod type.

@IDCs: I do have sims4 installed including a test setup so unless you really want this issue I think it makes sense I do it myself.

**Comment by IDCs** [comment]
@TanninOne, feel free to take this one as I don't have a Sims 4 test environment set-up :)

**Comment by fantesykikachu** [comment]
First @IDCs WickedWhims works fine in top level domain, what doesn't work is the batch file that the author ships with it, but all that does is modify the resource.cfg, (it also removes old installations of itself).

Second @TanninOne the file resource.cfg is created when a user enables mods in game, (I don't what happens if you delete it though) however the default is only
Priority 500
PackedFile *.package
PackedFile */*.package
this allows for one sub-directory, however if you run the batch file that comes with WickedWhims it modifies it to what you had. It is also important to note that WickedWhims while probably the most used mod is still controversial in the community and there shouldn't be the expectation that users use it. So if vortex changes the resource.cfg itself that would probably be fine.

**Comment by TanninOne** [comment]
@fantesykikachu do you have samples of mods that are distributed as loose files?

Vortex is designed to extract mods from archives because usually they consist of multiple files.
I think I can change it so that if the file can't be extracted as an archive and doesn't have an archive extension we offer to just take that file and put it into its own mod. But that works only as long as it's a single file or rather: each file you install like that would show up as its own mod.

**Comment by fantesykikachu** [comment]
almost all of the files on thesimsresource.com are distributed as loose .package files however these mods are only one file anyway; sites like http://modthesims.info/, http://sims4downloads.net, http://sims4updates.net/, and https://www.simsday.com/ typically redirect the creators tumblr or blogspot and most of them use loose .package files, even sites like LoversLab and nsfwmods while most files are compress some are not. 
and just for reference thesimsresource is the largest collection of mods for the sims 4, with over 100,000 mods it dwarfs the number skyrim mods on the nexus. (this is because most sims 4 mods are very small)

**Comment by TanninOne** [comment]
Please try out release 0.18.13 (https://github.com/Nexus-Mods/Vortex/releases/tag/v0.18.13), it should fix the issues you raised.

This version installs mods into a "Vortex Mods" subdirectory below the regular Mods directory so it doesn't interfere with mods you install manually. It should automatically clean up mods deployed through older versions of Vortex, please let us know if that worked for you.

**Comment by fantesykikachu** [comment]
So I tested 0.18.13 and both of the added features work, and the support for tray items is great, but I don't understand the purpose of the "Vortex Mods" subdirectory, the whole problem was around the default resource.cfg only allowing one subdirectory, which is now taken by the new "Vortex Mods" subdirectory so now with the default resource.cfg any mod that adds even a single subdirectory doesn't get loaded by the game.

also file name collisions aren't a problem because the game will overwrite it with whatever it loads last (that has the same name) I am not 100% sure of the order that it searches but base on a few mod patches installation guides I believe that it is alphabetical order lowest directory allowed by the  resource.cfg first.

I think the whole "Vortex Mods" subdirectory looks bad, and would honestly prefer if it were removed. If vortex could modify the resource.cfg itself ( to be like how wickedwhims does) that would probably be the easiest solution. If you could remove the "Vortex Mods" subdirectory, and let me know about modifying the resource.cfg then I will mark the first issue as complete ( I have also marked the second issue as complete).

With regards to the third issue I understand that expecting compressed mods is core to the design of vortex, so if you don't want to change that just let me know and i can mark that issue as either completed or as won't fix (you choose as a lot of people get mad at developers when the tern won't fix gets used). and with the last issue I have changed my mind and don't think we should support bad mod behavior, as any script mod that is up to date should not have any .pyo and should be in a .ts4script, but whether you want to add it is up to you but I will probably close this issue either way.

**Comment by TanninOne** [comment]
Unless there is a bug you are mistaken about the Vortex Mods directory.
The point of it is this: Vortex can now update the resource.cfg file and add its own block regarding "Vortex Mods" without changing the rest of the resource.cfg file.
This means you can have whatever resources.cfg setup you want, vortex will not break it but add to it such that the vortex-installed mods work.
You could even set up your own mod sub-directories for manually installed mods that would take precedence over the vortex-installed mods.

tl;dr: All mods you install through Vortex should just work (tm) and everything else should be unaffected by Vortex.

> I am not 100% sure of the order that it searches but base on a few mod patches installation guides I believe that it is alphabetical order lowest directory allowed by the resource.cfg first.

That's what the Priority lines in the resource.cfg are there for. The "default" mod directory has priority of 500, Vortex mods get a priority of 1337 and should therefore take precedence over the default mod directories.
If you set up further mod subdirectories with priority - say - 2000 - those would be loaded after the Vortex Mods.

> I think the whole "Vortex Mods" subdirectory looks bad, and would honestly prefer if it were removed. 

It's a lot cleaner this way though, as I described above.
The problem is that searching around on the internet, there are a lot of different suggestions for how the resource.cfg file should be set up. If Vortex just crowbars in its own resource.cfg, that may break existing setups. Having Vortex as a separate section is a lot safer.

> With regards to the third issue

This is actually also adressed in this release.
If you "install from file" or drag&drop a ".package" or '.ts4script' file for example in Vortex it should set it up as its own mod.
The same should technically also work for .py files or anything else.

**Comment by fantesykikachu** [comment]
Well the new directory isn't ideal for me but I realize that being an advanced user that uses it in non standard, means that I am not the target audience for this change and that it will work for most people, so I will mark it as complete.
 I noticed the "install from file" option but couldn't remember if that was in old releases, so I will mark that as complete.
and I guess I didn't explain the final issue properly, the game will not load a .py or .pyc file in the mods folder, it must be inside a .zip (or .ts4script) file inside the mods folder (it also won't load any .package files in the .zip file, those must be extracted).
now that the top three issues are complete I will probably close this issue tomorrow.

**Comment by TanninOne** [comment]
I still don't really understand why the "Vortex  Mods" subdirectory bothers you so much but if you say you can live with it I'll take that. ;)

Regarding the .py files: Yeah, don't think I'll be fixing that then. Keeping non-archives as they are (.package or .ts4script) is one thing but for Vortex to zip files automatically would be quite complex and it's such a niche use case I don't think it's worth it, at least not unless another couple of games show up that need something similar.

**Comment by fantesykikachu** [comment]
I apologize for reopening however it turns out the issue is not fixed. Because I was compressing all of my mod I didn't realize the problem that was created, as it happen while the game uses the resource.cfg to look for .package files, the pattern for .ts4scripts are hard-coded as *.ts4script or */*.ts4script in the executable, meaning a file like Mods/Vortex Mods/mod name/mod name.ts4script won't get loaded (however Mods/Vortex Mods/mod name.ts4script is fine).

**Comment by TanninOne** [comment]
Do you have an example of a mod that doesn't install correctly? (link please)

**Comment by fantesykikachu** [comment]
sorry for being so long, was busy and forgot, the two that I know of are wickedwhims (if you don't used the installer which you really shouldn't need to) and Nisa’s Wicked Perversions (https://www.loverslab.com/files/file/5002-nisa%E2%80%99s-wicked-perversions), both of these are on LL, but there are also lots mods which ship any dependencies with them and if any of them are scripts they would break (sorry can't think of one off the top of my head that includes a script like that.

**Comment by TanninOne** [comment]
Thank you, one example should be good enough.
Sorry I edited your post to make the link not-clickable but LL dances a bit too close to illegal territory for my taste.

**Comment by EustaceCS** [comment]
https://sacrificialmods.com/downloads.html
Road to Fame, Zombie Apocalypse, Life's Drama, etc.
(in case of necessity in non-adult popular Sims 4 mods which became even more broken after installing through mentioned Vortex 0.18.13 release)
Cheers!

**Comment by TanninOne** [comment]
@EustaceCS Can't reproduce. Both Road to Fame and Life's Drama seem to work perfectly fine installed through Vortex.

**Comment by TanninOne** [comment]
fixed by a1ca67767521360830061bcf81a6d2c362c2852f

With this fix ts4script and package files should now all be placed at the "Mods/Vortex Mods" level where they do work.
